src/Build/Instance/TaskRegistry.cs

```diff
@@ -682,37 +646,39 @@ private static bool IsTaskFactoryClass(Type type, object unused)
                 string[] nameComponents = taskName.Split('.');
                 string unqualifiedTaskName = nameComponents[nameComponents.Length - 1];
 
-                // Is the task already registered?
-                if (overriddenTasks.TryGetValue(unqualifiedTaskName, out List<RegisteredTaskRecord> recs))
+                List<RegisteredTaskRecord> records = _overriddenTasks.GetOrAdd(unqualifiedTaskName, _ => new List<RegisteredTaskRecord>());
+
+                lock (records)
                 {
-                    foreach (RegisteredTaskRecord rec in recs)
+                    if (records.Count == 0)
                     {
-                        if (rec.RegisteredName.Equals(taskIdentity.Name, StringComparison.OrdinalIgnoreCase))
-                        {
-                            loggingService.LogError(context, null, new BuildEventFileInfo(projectUsingTaskInXml.OverrideLocation), "DuplicateOverrideUsingTaskElement", taskName);
-                            break;
-                        }
+                        // New record's name may be fully qualified. Use it anyway to account for partial matches.
+                        records.Add(newRecord);
+                        loggingService.LogComment(context, MessageImportance.Low, "OverrideUsingTaskElementCreated", taskName, projectUsingTaskInXml.OverrideLocation);
+                    }
+                    // Is the task already registered?
+                    else if (records.Any(rec => rec.RegisteredName.Equals(taskIdentity.Name, StringComparison.OrdinalIgnoreCase)))
+                    {
+                        loggingService.LogError(context, null, new BuildEventFileInfo(projectUsingTaskInXml.OverrideLocation), "DuplicateOverrideUsingTaskElement", taskName);
+                    }
+                    else
+                    {
+                        records.Add(newRecord);
                     }
-                    recs.Add(newRecord);
-                }
-                else
-                {
-                    // New record's name may be fully qualified. Use it anyway to account for partial matches.
-                    List<RegisteredTaskRecord> unqualifiedTaskNameMatches = new();
-                    unqualifiedTaskNameMatches.Add(newRecord);
-                    overriddenTasks.Add(unqualifiedTaskName, unqualifiedTaskNameMatches);
-                    loggingService.LogComment(context, MessageImportance.Low, "OverrideUsingTaskElementCreated", taskName, projectUsingTaskInXml.OverrideLocation);
                 }
             }
 
-            registeredTaskEntries.Add(newRecord);
+            lock (registeredTaskEntries)
+            {
+                registeredTaskEntries.Add(newRecord);
```

Comment:
Is it really correct to add the new record without checking if it's already there? (Think two threads running this method in parallel.)

Reply:
Technically it would be fine - as we allways select just one task. Question is the perf impact

Looking on the execution paths leading to here (either from `Evaluate` or from `ExecuteTarget`), there is currently no option of executing this method concurrently - so I think even the `lock` is actually wrong here - so removing that one.

Does that sound ok?

Reply:
As long as this is provably separate from the code paths where we're currently hitting concurrency issues (which, presumably led us to this PR), it sounds OK!

Reply:
On a second thought (and crawlingthrough the code) - I'm not so 100% sure that the writer part cannot be concurrent (and it can always be attempted to be made concrrent in the future) - so rather keeping the lock.

As for double registration - that sounds as a niche case, that would force all registrations to first traverse existing lists (despite they'll most often have zero items), so I believe the perf tradeoff is correct here. But I'm open to opposing ideas :-)